### PR TITLE
6.0.0-beta.1

### PR DESCRIPTION
**Pull request by aembke**
* Add DNS interface
* Remove TLS env flags
* Improve connectivity checks on writes
* Update docs and examples

### CODE REVIEW

File: src/interfaces.rs [hunk 1]
```diff
@@ -153,6 +161,13 @@ pub trait ClientLike: Clone + Send + Sized {
     self.inner().server_state.read().server_version()
   }
 
+  /// Override the DNS resolution logic for the client.
+  #[cfg(feature = "dns")]
+  #[cfg_attr(docsrs, doc(cfg(feature = "dns")))]
+  async fn set_resolver(&self, resolver: Arc<dyn Resolve>) {
```
Comment (DDtKey) [inline src/interfaces.rs:167]: Could you reflect this function to the `Pool` implementation? It should update all underlying clients 🤔

Comment (aembke) [reply src/interfaces.rs:167]: Good catch, yes definitely.

**Comment by DDtKey** [comment]
Access by index ([here](https://github.com/aembke/fred.rs/blob/2cc2ecc4d2e7f7a86c16042c4d602b276bb7bff9/src/protocol/responders.rs#L237)) could panic and stop `tokio` thread
It happened in CI: [link](https://app.circleci.com/pipelines/github/aembke/fred.rs/168/workflows/19277eaf-f7cc-4477-94b9-00c108b10fab/jobs/1232/parallel-runs/0/steps/0-112) 🤔

**Comment by DDtKey** [comment]
A lot of work has been done here and for `6.0` release in general. Thank you @aembke!

Now it looks much more reliable. 
I gonna perform some my own failover  tests under the load with these changes today, but simple tests are working and looks like the issues were covered! 🚀

**Comment by DDtKey** [comment]
Hey @aembke! 
Some of my observation with current state of PR:

In case one of re-connection (clustered) I'm able to face with the following logs:
```
fred-2HamTxultz: Recv CLUSTER SLOTS response: Null
fred-2HamTxultz: Rebuilding cluster state from host: localhost
fred-2HamTxultz: Changing cluster connections: ClusterChange { add: [], remove: [] }
``` 

So looks like there is no correct handling of such situations, when CLUSTER SLOTS slots returned `Null` for some reason 🤔 

I also thought, if it's correctly to trust any response from the node? 
Because node could became available before it's synced with cluster state.

I tried to check results of redis `CLUSTER INFO` in case if one of the node was restarted.
Just take a look at the example with `redis-cli` (cluster of 3 masters, without replicas):

1. **`CLUSTER INFO` when cluster works fine** (state - **ok**, size = **3**)
```
redis-node-2:6379> CLUSTER INFO
cluster_state:ok
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:3
cluster_size:3
# ... some fields are ommited
```

2. **`CLUSTER INFO` immediately after restarting** (state - **fail**, and size = **1**) 
```
redis-node-2:6379> CLUSTER INFO
cluster_state:fail
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:3
cluster_size:1
# ... some fields are ommited
```
**Question:** if we would reconnect to this node, we will update slots incorrectly?  I see no checks for this, so I guess so.


3. But after some time, cluster became synced of course, and it returns **state = ok, size = 3** again
```
redis-node-2:6379> CLUSTER INFO
cluster_state:ok
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:3
cluster_size:3
# ... some fields are ommited
```

Probably I'm missing something, but should we add some check for that? Because logs I've attached above looks weird

Let me know if I can help somehow, I can participate and send some patches if we would achieve understanding and agreement 🙂

**Comment by aembke** [comment]
Hey @DDtKey, thanks for diving into all that. 

Yeah, the way it works now is not great. If `CLUSTER SLOTS` returns `Null` it results in an empty routing table and connection map, but doesn't fail on that step. Later when the caller tries to send a command to the server it'll fail and trigger the reconnect logic, and presumably the cluster state would have resolved itself in that time. That's not a great solution so I'll change it to check the state via `CLUSTER INFO` like you note.

**Comment by aembke** [comment]
Hey @DDtKey, let's move this conversation to a new thread somewhere on here since I think you're touching on some broader design discussions around when it's appropriate to do some magic that makes things "just work."

In this specific case I'm going to make the `cluster_state:ok` check a required check for any connections to a clustered node. From what I can gather from the `CLUSTER INFO` docs that seems to be the intention with that command. In my opinion it's valid to say that a cluster node that is "up" but unable to accept writes should be considered "down" generally. 

I'll be adding a replica feature in the coming weeks but that will be explicitly opt-in for callers. There are consistency trade offs with replicas and I do not want to use them unless the caller is very aware that they're being used.

**Comment by DDtKey** [comment]
Yes, sure, thanks!

**Pull request merged**